src/main/java/io/r2dbc/mssql/LoginFlow.java

```diff
@@ -99,8 +101,27 @@ private LoginFlow() {
                     }
 
                     if (DoneToken.isDone(message)) {
-                        sink.next(message);
+
+                        LoginExchangeResult result = Optional.of(routingData)
+                            .map(AtomicReference::get)
+                            .map(it -> LoginExchangeResult.routed(it.getServerName(),
+                                it.getPort()))
+                            .orElseGet(LoginExchangeResult::connected);
+
+                        sink.next(result);
                         sink.complete();
+
+                        return;
+                    }
+
+                    if (message instanceof EnvChangeToken) {
```

Comment:
We keep most items that are provided via `EnvChangeToken` within the client (e.g. transaction id, database, locale). How about moving `RoutingData` into `Client` and expose it as `Optional<RoutingData> getRedirect()`?

Reply:
I have reverted the changes in {{LoginFlow}} and added a listener on the client for the redirect instead. I think this is more in line with the existing pattern. Let me know what you think.